File: baekjoon/2580/2580.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

const boardLen int = 9

func promising(board [][]int, row, column int) bool {
	for i := 0; i < boardLen; i++ {
		if i != column && board[row][i] == board[row][column] {
			return false
		}

		if i != row && board[i][column] == board[row][column] {
			return false
		}
	}

	startRow := row - (row % 3)
	startColumn := column - (column % 3)
	for i := startRow; i <= startRow+2; i++ {
		for j := startColumn; j <= startColumn+2; j++ {
			if i != row && j != column && board[i][j] == board[row][column] {
				return false
			}
		}
	}

	return true
}

func dfs(board [][]int, xInfo, yInfo []int, count int, find *bool) {
	if count == len(xInfo) {
		*find = true

		for i := 0; i < boardLen; i++ {
			for j := 0; j < boardLen; j++ {
				fmt.Printf("%d ", board[i][j])
			}
			fmt.Println()
		}

		return
	}

	for i := 1; i <= boardLen; i++ {
		board[xInfo[count]][yInfo[count]] = i
		if promising(board, xInfo[count], yInfo[count]) {
			dfs(board, xInfo, yInfo, count+1, find)

			if *find {
				return
			}
		}
		board[xInfo[count]][yInfo[count]] = 0
	}
}

func main() {
	reader := bufio.NewReader(os.Stdin)
	writer := bufio.NewWriter(os.Stdout)
	defer writer.Flush()

	board := make([][]int, boardLen)
	for i := 0; i < boardLen; i++ {
		board[i] = make([]int, boardLen)
	}

	xInfo := make([]int, 0)
	yInfo := make([]int, 0)

	for i := 0; i < boardLen; i++ {
		for j := 0; j < boardLen; j++ {
			fmt.Fscanf(reader, "%d", &board[i][j])

			if board[i][j] == 0 {
				xInfo = append(xInfo, i)
				yInfo = append(yInfo, j)
			}
		}
		fmt.Fscanln(reader)
	}

	find := false
	dfs(board, xInfo, yInfo, 0, &find)
}

File: baekjoon/2580/2580.py
import sys


class Sudoku:

    def __init__(self):
        self.find = False

        self.board = []
        for i in range(9):
            self.board.append(list(map(int, sys.stdin.readline().split())))

        self.zeroX = {}
        self.zeroY = {}
        self.zeroCount = 0
        for i in range(9):
            for j in range(9):
                if self.board[i][j] == 0:
                    self.zeroX[self.zeroCount] = i
                    self.zeroY[self.zeroCount] = j
                    self.zeroCount += 1

    def promising(self, row, column):
        for i in range(9):
            if i != column and self.board[row][i] == self.board[row][column]:
                return False
            if i != row and self.board[i][column] == self.board[row][column]:
                return False

        x = row - row % 3
        y = column - column % 3
        for i in range(x, x + 3):
            for j in range(y, y + 3):
                if i != row and j != column and self.board[i][j] == self.board[
                        row][column]:
                    return False

        return True

    def dfs(self, count):
        if count == self.zeroCount:
            self.find = True
            return

        for i in range(1, 10):
            self.board[self.zeroX[count]][self.zeroY[count]] = i
            if self.promising(self.zeroX[count], self.zeroY[count]):
                self.dfs(count + 1)
                if self.find:
                    return
            self.board[self.zeroX[count]][self.zeroY[count]] = 0

    def run(self):
        self.dfs(0)

        for i in range(9):
            for j in range(9):
                print(self.board[i][j], end=' ')
            print()


def main():
    Sudoku().run()


if __name__ == "__main__":
    main()
